src/lib/export/export.ts

```diff
@@ -0,0 +1,209 @@
+import fs from 'fs'
+import path from 'path'
+import crypto from 'crypto'
+import remote from 'electron'
+import { DBConnection, TableOrView, TableFilter, TableResult } from '@/lib/db/client'
+
+export interface ExportOptions {
+  chunkSize: number,
+  deleteOnAbort: boolean
+}
+
+export abstract class Export {
+  id: string
+  connection: DBConnection
+  countExported: number = 0
+  countTotal: number = 0
+  error: Error | null = null
+  filePath: string
+  fileSize: number = 0
+  filters: TableFilter[] | any[]
+  lastChunkTime: number = 0
+  options: ExportOptions
+  outputOptions: any
+  showNotification: boolean = true
+  status: Export.Status = Export.Status.Idle
+  table: TableOrView
+  timeElapsed: number = 0
+  timeLeft: number = 0
+
+  abstract getHeader(firstRow: any): Promise<string | void>
+  abstract getFooter(): Promise<string | void>
+  abstract formatChunk(data: any): string[]
+
+  constructor(
+    filePath: string,
+    connection: DBConnection,
+    table: TableOrView,
+    filters: TableFilter[] | any[],
+    options: ExportOptions,
+    outputOptions: any
+  ) {
+    this.filePath = filePath
+    this.connection = connection
+    this.table = table
+    this.filters = filters
+    this.options = options
+    this.outputOptions = outputOptions
+    this.id = this.generateId()
+  }
+
+  generateId(): string {
+    const md5sum = crypto.createHash('md5')
+
+    md5sum.update(Date.now().toString(), 'utf8')
+    md5sum.update(this.table.name)
+    md5sum.update(this.filePath)
+
+    return md5sum.digest('hex')
+  }
+
+  async getChunk(offset: number, limit: number): Promise<TableResult | undefined> {
+    const result = await this.connection.selectTop(
```

Comment:
It would possibly be better to implement a separate `select` method that doesn't return the row count. However, when testing it the performance gain from not querying the row count was negligible.

Reply:
So three things to address here:

- I just added an updated rowCount implementation that relies on an approximate count rather than the real count, as on very large tables this is prohibitively expensive. So we do probably want to avoid that.

- Because the sequence of selectTop operations don't take place inside a single transaction there could be duplicates as it pages through the rows when new rows are inserted.
- Most database clients provide a way to maintain a pointer and page through the results as they stream from the database, we're best to use this in order to get a consistent snapshot.

The amount of work you have done already exceeds my expectations. Really really good stuff. So I'm happy taking on this backend code to make it 'do the right thing'.